Rebase with Upstream and Port to Python3

Saumya-Mishra9129 opened the pull request:
Twisted uses a module named 'attr' which I couldn't find in sugar . So I installed it manually using pip (apt-get didn't work) with following command-

`sudo pip install attrs `

Also we need to install OpenSSL with `sudo apt-get install python3-openssl` to get it working.

@chimosky @quozl @srevinsaju Please Review.

Saumya-Mishra9129 committed:
change exec to sugar-activity3

Saumya-Mishra9129 committed:
Run 2to3

Saumya-Mishra9129 committed:
Rebase with upstream constantly

constantly source rebasing with https://github.com/twisted/constantly

Saumya-Mishra9129 committed:
Rebase with upstream : incremental

incremental source rebasing with https://github.com/twisted/incremental

Saumya-Mishra9129 committed:
Rebase with upstream : twisted

twisted source rebasing with https://github.com/twisted/twisted

Saumya-Mishra9129 committed:
Rebase with upstream : Zope

Zope source rebasing with https://github.com/zopefoundation/zope.interface

Saumya-Mishra9129 committed:
Change shebang to python3

Saumya-Mishra9129 committed:
TypeError : Item 0 : must be number not str

Saumya-Mishra9129 committed:
TypeError: Item 0: must be number not str

srevinsaju commented on afk_manager.py:
```diff
@@ -35,20 +35,20 @@ def __init__(self):
-        if not nickname in self.timeout_ids.keys() or restart:
-            if nickname in self.timeout_ids.keys():
+        if not nickname in list(self.timeout_ids.keys()) or restart:
+            if nickname in list(self.timeout_ids.keys()):
```
`list` function can be ignored. I assume these were created by `2to3`. 
As mentioned [here](https://github.com/sugarlabs/sugar-docs/blob/master/src/python-porting-guide.md),
> Port from Python 2 to Python 3. Start your porting with 2to3 tool,
> In the terminal, type:
> `2to3 -w -n *.py`
> and then **_review_ every change made,**

Saumya-Mishra9129 replied:
Yeah it is created by 2to3. I will fix that

srevinsaju commented on afk_manager.py:
```diff
@@ -35,20 +35,20 @@ def __init__(self):
 
     def remove_nickname(self, nickname):
-        if nickname in self.timeout_ids.keys():
+        if nickname in list(self.timeout_ids.keys()):
```
same

srevinsaju commented on chat_box.py:
```diff
@@ -307,7 +307,7 @@ def set_topic(self, channel, topic):
 
     def remove_nickname_from_all_channels(self, nickname):
-        for channel in self.nicks.keys():
+        for channel in list(self.nicks.keys()):
```
same

srevinsaju commented on chat_box.py:
```diff
@@ -319,7 +319,7 @@ def _change_topic(self, widget, topic):
 
     def set_user_afk(self, nickname, afk):
-        for channel in self.nicks.keys():
+        for channel in list(self.nicks.keys()):
```
here also

srevinsaju commented on constantly/_constants.py:
```diff
@@ -167,7 +167,7 @@ def __new__(self, name, bases, attributes):
 
         constants = []
-        for (name, descriptor) in attributes.items():
+        for (name, descriptor) in list(attributes.items()):
```
This can be made more readable by
```python
>>> a = {'a':"apple", 'b':"ball", 'c':"cat"}
>>> for name, descriptor in a.items():
...     print(name, descriptor)
... 
a apple
b ball
c cat
>>> 

```

Notice the extra parenthesis has been removed; and list function also has been removed

srevinsaju reviewed (changes requested):
Reviewing is tedious, 555 files from the last two commits. Some minor corrections; Will update you with sugar testing

Saumya-Mishra9129 committed:
Run 2to3

chimosky commented:
@Saumya-Mishra9129 said
> Yeah it is probably a good thing to be done. If others agree I will do that as well.

Yeah I agree, use distro specific packages where possible. 

In `setup.py` `cmd_genpot` add `--from-code=UTF-8` as part of your args to `xgettext` as an error is thrown by `xgettext` when trying to parse some strings. See https://github.com/sugarlabs/sugar-toolkit-gtk3/commit/ecf335c19cb80471dcda8d2481382bf5d34c6a5f.

Didn't review cd8a53e...5814bac as I can't properly view the diff of the aggregate change.
I'll wait for you to use distro specific packages before reviewing.

Saumya-Mishra9129 committed:
Remove list function imported by 2to3

Saumya-Mishra9129 committed:
Some PEP8 fixes

Saumya-Mishra9129 committed:
Optimize imports

Saumya-Mishra9129 committed:
Remove zope, constantly, twisted, incremental

support the use of distro specific packages

Saumya-Mishra9129 committed:
Add --from-code parameter to xgettext for a python3 activity

Based on https://github.com/sugarlabs/sugar-toolkit-gtk3/commit/ecf335c19cb80471dcda8d2481382bf5d34c6a5f

Saumya-Mishra9129 committed:
Update Readme.md with Python3 commands and add python3-twisted

Saumya-Mishra9129 commented:
> Didn't review cd8a53e...5814bac as I can't properly view the diff of the aggregate change.
I'll wait for you to use distro specific packages before reviewing.

It is done. Please review.

quozl commented:
What do you need for the UnicodeDecodeError?  Are all the source files UTF-8 encoded, and do they contain only UTF-8?

Saumya-Mishra9129 committed:
Fix UnicodeDecodeError

Fixes error seen while making connection.
while sending a line , it takes channel as an encoded object.

Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/twisted/python/log.py", line 103, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/usr/lib/python3/dist-packages/twisted/python/log.py", line 86, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/usr/lib/python3/dist-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/lib/python3/dist-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/usr/lib/python3/dist-packages/twisted/internet/posixbase.py", line 614, in _doReadOrWrite
    why = selectable.doRead()
  File "/usr/lib/python3/dist-packages/twisted/internet/tcp.py", line 243, in doRead
    return self._dataReceived(data)
  File "/usr/lib/python3/dist-packages/twisted/internet/tcp.py", line 249, in _dataReceived
    rval = self.protocol.dataReceived(data)
  File "/usr/lib/python3/dist-packages/twisted/words/protocols/irc.py", line 2631, in dataReceived
    basic.LineReceiver.dataReceived(self, data)
  File "/usr/lib/python3/dist-packages/twisted/protocols/basic.py", line 572, in dataReceived
    why = self.lineReceived(line)
  File "/usr/lib/python3/dist-packages/twisted/words/protocols/irc.py", line 2637, in lineReceived
    line = line.decode("utf-8")
builtins.UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe4 in position 121: invalid continuation byte

Tested with Ubuntu 20.04 and sucrose version 0.117

Saumya-Mishra9129 commented:
Update - UnicodeDecodeError is fixed in d4b7c80. I have tested also works fine. @chimosky @srevinsaju @quozl Please Review.

quozl commented:
Thanks.

Fixes https://github.com/sugarlabs/PolariActivity/issues/6

Polari upstream has changed from Python to JavaScript.  But this code remains Python.  When you say you rebased with upstream, what did you do?

Reviewed.  Prepared a diffstat.  Most of the change is to remove the embedded twisted, which is good.

Tested.  Speaking on #sugar without registration gives no message.  A message is sent to the client by the server but is not displayed.

Saumya-Mishra9129 commented:
> Polari upstream has changed from Python to JavaScript. But this code remains Python. When you say you rebased with upstream, what did you do?

By rebasing with upstream I thought of rebasing libraries used ( twisted , zope, incremental and constantly) with their upstream sources. So I first did that, but then @chimosky suggested to use distro specific twisted ,which I have done later.

quozl commented:
I see, thanks.  Our toolkit doesn't have support for JavaScript GObject Introspection apps, so rebasing Polari itself would be a major undertaking requiring new toolkit code as well.  I don't think we need that yet.  But you might check for changes to Polari between the time this source code was copied and before _they_ switched to JavaScript.

Saumya-Mishra9129 commented:
> I see, thanks. Our toolkit doesn't have support for JavaScript GObject Introspection apps, so rebasing Polari itself would be a major undertaking requiring new toolkit code as well. I don't think we need that yet. But you might check for changes to Polari between the time this source code was copied and before _they_ switched to JavaScript.
> 

Thanks, for Polari source code, I couldn't find source code which was related to Python3 and also that's so old also.

>That could be a separate pull request though.

Thanks I will make a different one. Until then you can merge these changes.

quozl commented:
I've checked again, and I no longer think this activity has any association with the GNOME Polari app.  Polari activity is an independent implementation using Twisted, and the name is coincidental.
